proxy/proxy.go

```diff
@@ -527,3 +557,47 @@ func (proxy *Proxy) updateContainerNetworkSettings(container jsonObject) error {
 	networkSettings["IPPrefixLen"], _ = nets[0].Mask.Size()
 	return nil
 }
+
+// containerAttachLoop manages container attachment retry timers, to ensure
+// that failed containers are retried. new containers should be sent to
+// proxy.attachContainer
+func (proxy *Proxy) containerAttachLoop() {
+	defer proxy.wg.Done()
+	var (
+		a              attachJob
+		nextRetry      <-chan time.Time
+		pendingRetries []attachJob
+	)
+	now := time.Now()
+	for {
+		if len(pendingRetries) > 0 && nextRetry == nil {
+			nextRetry = time.After(pendingRetries[0].tryAfter.Sub(now))
+		}
+		select {
+		case <-nextRetry:
+			nextRetry = nil
+			a = pendingRetries[0]
+			pendingRetries = pendingRetries[1:]
+		case a = <-proxy.attachContainer:
+			for i, j := range pendingRetries { // remove any existing entry for same id
```

Comment:
You mean "the entry which was previously scheduled to be tried next"? Yes. It seemed innocuous to try the next one at that time.

It seems that `pendingRetries` could be left empty, which would be bad.

Perhaps best to set `nextRetry` to nil if an entry is taken out of the list, so it is re-computed next time round the loop.